flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/DispatcherResourceCleanupTest.java

```diff
@@ -651,6 +661,104 @@ public void testFailingJobManagerRunnerCleanup() throws Exception {
         awaitStatus(dispatcherGateway, jobId, JobStatus.RUNNING);
     }
 
+    @Test(timeout = 5000L)
+    public void testArchiveSuccessfullyWithinTimeout() throws Exception {
+
+        final Configuration configuration = new Configuration();
+        configuration.setLong(ClusterOptions.CLUSTER_SERVICES_SHUTDOWN_TIMEOUT, 1000L);
+
+        final ExecutorService ioExecutor = Executors.newSingleThreadExecutor();
+
+        try {
+            final TestingHistoryServerArchivist archivist =
+                    new TestingHistoryServerArchivist(ioExecutor, 50L);
+            final TestingJobMasterServiceLeadershipRunnerFactory testingJobManagerRunnerFactory =
+                    new TestingJobMasterServiceLeadershipRunnerFactory(0);
+            final TestingDispatcher.Builder testingDispatcherBuilder =
+                    createTestingDispatcherBuilder()
+                            .setHistoryServerArchivist(archivist)
+                            .setConfiguration(configuration);
+            startDispatcher(testingDispatcherBuilder, testingJobManagerRunnerFactory);
+
+            submitJobAndWait();
+            final TestingJobManagerRunner testingJobManagerRunner =
+                    testingJobManagerRunnerFactory.takeCreatedJobManagerRunner();
+            finishJob(testingJobManagerRunner);
+
+            globalCleanupFuture.join();
+            dispatcher.getJobTerminationFuture(jobId, Time.milliseconds(1000L)).join();
+
+            assertTrue(archivist.isArchived());
```

Comment:
If we use a CompletableFuture instead of sleep, we would check, that the cleanup wasn't triggered and the job didn't terminate before completing that future here. Then, completing the future should result in the cleanup being triggered and job finally terminating...

Reply:
Thank you for providing a better solution here. I've replaced the `Thread.sleep` with the `CompletableFuture`.

Reply:
We could simplify it even more. We don't need the `TestingHistoryServerArchivist` class but just return a callback `executionGraphInfo -> resultFuture`. The test itself would look like this, then:
```
final CompletableFuture<Acknowledge> archiveFuture = new CompletableFuture<>();
final TestingJobManagerRunner testingJobManagerRunner =
startDispatcherAndSubmitJob(archiveTimeout, executionGraphInfo -> archiveFuture);
```

You can simulate the completion of the archiving by completing `archiveFuture` like you did already in `testArchiveSuccessfullyWithinTimeout`